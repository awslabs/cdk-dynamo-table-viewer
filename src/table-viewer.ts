import * as path from 'path';
import {
  aws_apigateway as apigw,
  aws_lambda as lambda,
  aws_dynamodb as dynamodb,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface TableViewerProps {
  /**
   * The endpoint type of the
   * [LambdaRestApi](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-apigateway.LambdaRestApi.html)
   * that will be created
   * @default - EDGE
   */
  readonly endpointType?: apigw.EndpointType;

  /**
   * The DynamoDB table to view. Note that all contents of this table will be
   * visible to the public.
   */
  readonly table: dynamodb.Table;

  /**
   * The web page title.
   * @default - No title
   */
  readonly title?: string;

  /**
   * Name of the column to sort by, prefix with "-" for descending order.
   * @default - No sort
   */
  readonly sortBy?: string;
}

/**
 * Installs an endpoint in your stack that allows users to view the contents
 * of a DynamoDB table through their browser.
 */
export class TableViewer extends Construct {

  public readonly endpoint: string;

  constructor(parent: Construct, id: string, props: TableViewerProps) {
    super(parent, id);

    const handler = new lambda.Function(this, 'Rendered', {
      code: lambda.Code.fromAsset(path.join(__dirname, '..', 'lambda')),
      runtime: lambda.Runtime.NODEJS_18_X,
      handler: 'index.handler',
      environment: {
        TABLE_NAME: props.table.tableName,
        TITLE: props.title || '',
        SORT_BY: props.sortBy || '',
      },
    });

    props.table.grantReadData(handler);

    const home = new apigw.LambdaRestApi(this, 'ViewerEndpoint', {
      handler,
      endpointConfiguration: props.endpointType
        ? { types: [props.endpointType] }
        : undefined,
    });
    this.endpoint = home.url;
  }
}
