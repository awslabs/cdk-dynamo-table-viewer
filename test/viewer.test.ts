import { Stack, App } from 'aws-cdk-lib';
import * as assertions from 'aws-cdk-lib/assertions';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import { TableViewer } from '../src';

test('happy  flow', () => {
  // GIVEN
  const app = new App();
  const stack = new Stack(app, 'test');
  const table = new dynamodb.Table(stack, 'MyTable', {
    partitionKey: { name: 'id', type: dynamodb.AttributeType.STRING },
  });

  // WHEN
  new TableViewer(stack, 'viewer', { table });

  // THEN
  const template = assertions.Template.fromStack(stack);

  // keeping this so that we can compare changes for v2 upgrade
  expect(template).toMatchSnapshot();

  template.hasResourceProperties('AWS::Lambda::Function', {
    Handler: 'index.handler',
    Environment: {
      Variables: {
        SORT_BY: '',
        TABLE_NAME: {
          Ref: 'MyTable794EDED1',
        },
        TITLE: '',
      },
    },
    Runtime: 'nodejs18.x',
  });
  template.hasResourceProperties('AWS::ApiGateway::RestApi', {
    Name: 'ViewerEndpoint',
  });
  template.hasResourceProperties('AWS::ApiGateway::Stage', {
    StageName: 'prod',
  });
  template.hasResourceProperties('AWS::ApiGateway::Resource', {
    PathPart: '{proxy+}',
  });

  template.resourceCountIs('AWS::ApiGateway::Method', 2);
  template.hasResourceProperties('AWS::ApiGateway::Method', {
    HttpMethod: 'ANY',
    Integration: {
      IntegrationHttpMethod: 'POST',
      Type: 'AWS_PROXY',
    },
  });
  template.hasResourceProperties('AWS::IAM::Policy', {
    PolicyName: 'viewerRenderedServiceRoleDefaultPolicy196964DF',
    PolicyDocument: {
      Statement: assertions.Match.arrayWith([
        {
          Action: [
            'dynamodb:BatchGetItem',
            'dynamodb:GetRecords',
            'dynamodb:GetShardIterator',
            'dynamodb:Query',
            'dynamodb:GetItem',
            'dynamodb:Scan',
            'dynamodb:ConditionCheckItem',
            'dynamodb:DescribeTable',
          ],
          Effect: 'Allow',
          Resource: assertions.Match.arrayWith([
            {
              'Fn::GetAtt': ['MyTable794EDED1', 'Arn'],
            },
          ]),
        },
      ]),
      Version: '2012-10-17',
    },
    Roles: [{
      Ref: 'viewerRenderedServiceRole7AE70B4C',
    }],
  });
  template.resourceCountIs('AWS::DynamoDB::Table', 1);
  template.resourceCountIs('AWS::ApiGateway::Account', 1);
  template.resourceCountIs('AWS::ApiGateway::Deployment', 1);
  template.resourceCountIs('AWS::IAM::Role', 2);
  template.resourceCountIs('AWS::Lambda::Permission', 4);
});
