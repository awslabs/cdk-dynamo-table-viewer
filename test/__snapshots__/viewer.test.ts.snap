// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`happy  flow 1`] = `
Object {
  "Outputs": Object {
    "viewerViewerEndpoint2EEAF9D2": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "viewerViewerEndpointDED2DEA4",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "viewerViewerEndpointDeploymentStageprod95943DC5",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "MyTable794EDED1": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "viewerRenderedEEFA1E5E": Object {
      "DependsOn": Array [
        "viewerRenderedServiceRoleDefaultPolicy196964DF",
        "viewerRenderedServiceRole7AE70B4C",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "6ef63c8e0feb2ee73d336465b0d50733c8ede8bfdb37d27af4a6582fecd14ed0.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "SORT_BY": "",
            "TABLE_NAME": Object {
              "Ref": "MyTable794EDED1",
            },
            "TITLE": "",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "viewerRenderedServiceRole7AE70B4C",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "viewerRenderedServiceRole7AE70B4C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "viewerRenderedServiceRoleDefaultPolicy196964DF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "MyTable794EDED1",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "viewerRenderedServiceRoleDefaultPolicy196964DF",
        "Roles": Array [
          Object {
            "Ref": "viewerRenderedServiceRole7AE70B4C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "viewerViewerEndpointANY4A9252D2": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "viewerRenderedEEFA1E5E",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "viewerViewerEndpointDED2DEA4",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "viewerViewerEndpointDED2DEA4",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "viewerViewerEndpointANYApiPermissionTesttestviewerViewerEndpointDCCECA7FANYF9156524": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "viewerRenderedEEFA1E5E",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "viewerViewerEndpointDED2DEA4",
              },
              "/test-invoke-stage/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "viewerViewerEndpointANYApiPermissiontestviewerViewerEndpointDCCECA7FANY731E84A4": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "viewerRenderedEEFA1E5E",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "viewerViewerEndpointDED2DEA4",
              },
              "/",
              Object {
                "Ref": "viewerViewerEndpointDeploymentStageprod95943DC5",
              },
              "/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "viewerViewerEndpointAccount1D965397": Object {
      "DeletionPolicy": "Retain",
      "DependsOn": Array [
        "viewerViewerEndpointDED2DEA4",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "viewerViewerEndpointCloudWatchRole28045685",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "viewerViewerEndpointCloudWatchRole28045685": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "viewerViewerEndpointDED2DEA4": Object {
      "Properties": Object {
        "Name": "ViewerEndpoint",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "viewerViewerEndpointDeployment00B069459b597739070db86e729796710f9f86da": Object {
      "DependsOn": Array [
        "viewerViewerEndpointproxyANYF7759A2A",
        "viewerViewerEndpointproxy103C52F5",
        "viewerViewerEndpointANY4A9252D2",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "viewerViewerEndpointDED2DEA4",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "viewerViewerEndpointDeploymentStageprod95943DC5": Object {
      "DependsOn": Array [
        "viewerViewerEndpointAccount1D965397",
      ],
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "viewerViewerEndpointDeployment00B069459b597739070db86e729796710f9f86da",
        },
        "RestApiId": Object {
          "Ref": "viewerViewerEndpointDED2DEA4",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "viewerViewerEndpointproxy103C52F5": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "viewerViewerEndpointDED2DEA4",
            "RootResourceId",
          ],
        },
        "PathPart": "{proxy+}",
        "RestApiId": Object {
          "Ref": "viewerViewerEndpointDED2DEA4",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "viewerViewerEndpointproxyANYApiPermissionTesttestviewerViewerEndpointDCCECA7FANYproxyC67CDEE2": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "viewerRenderedEEFA1E5E",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "viewerViewerEndpointDED2DEA4",
              },
              "/test-invoke-stage/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "viewerViewerEndpointproxyANYApiPermissiontestviewerViewerEndpointDCCECA7FANYproxy04819798": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "viewerRenderedEEFA1E5E",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "viewerViewerEndpointDED2DEA4",
              },
              "/",
              Object {
                "Ref": "viewerViewerEndpointDeploymentStageprod95943DC5",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "viewerViewerEndpointproxyANYF7759A2A": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "viewerRenderedEEFA1E5E",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "viewerViewerEndpointproxy103C52F5",
        },
        "RestApiId": Object {
          "Ref": "viewerViewerEndpointDED2DEA4",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
